### PR TITLE
pr for noaa branch

### LINKED ISSUE #98: Installation issue: esmf-8.3.0 depends on cmake, previous versions didn't
### Steps to reproduce the issue

Use `spack-stack-1.0.0-rc1` to install `esmf@8.3.0` on a system that does not have a default `cmake` in the search path

### Error message

```
...
dir=`pwd`; cd /home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/mod/modO/Linux.gfortran.64.mpich2.default; make -f ${dir}/makefile MAKEFILE=${dir}/makefile esmflib
make[7]: Entering directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/mod/modO/Linux.gfortran.64.mpich2.default'
make[7]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/mod/modO/Linux.gfortran.64.mpich2.default'
make[6]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/ArrayBundle/doc'
make[5]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/ArrayBundle'
tree_lib in: /home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/IO
make[5]: Entering directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/IO'
dir=`pwd`; cd /home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/mod/modO/Linux.gfortran.64.mpich2.default; make -f ${dir}/makefile MAKEFILE=${dir}/makefile esmflib
make[6]: Entering directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/mod/modO/Linux.gfortran.64.mpich2.default'
make[6]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/mod/modO/Linux.gfortran.64.mpich2.default'
tree_lib in: /home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/IO/PIO
make[6]: Entering directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/IO/PIO'
makefile:41: warning: overriding recipe for target 'tree_lib'
/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:2278: warning: ignoring old recipe for target 'tree_lib'
cmake --version
make[6]: cmake: Command not found
make[6]: *** [makefile:41: tree_lib] Error 127
make[6]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/IO/PIO'
make[5]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:3770: tree] Error 1
make[5]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/IO'
make[4]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:3770: tree] Error 1
make[4]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure'
make[3]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:3770: tree] Error 1
make[3]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src'
make[2]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:3770: tree] Error 1
make[2]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src'
make[1]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:2245: build_libs] Error 2
make[1]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src'
make: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:2237: lib] Error 2
==> Error: ProcessError: Command exited with status 2:
    'make' '-j8'

7 errors found in build log:
     2515    tree_lib in: /home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructure/IO/PIO
     2516    make[6]: Entering directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructu
             re/IO/PIO'
     2517    makefile:41: warning: overriding recipe for target 'tree_lib'
     2518    /home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:2278: warning: ignoring old r
             ecipe for target 'tree_lib'
     2519    cmake --version
     2520    make[6]: cmake: Command not found
  >> 2521    make[6]: *** [makefile:41: tree_lib] Error 127
     2522    make[6]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructur
             e/IO/PIO'
  >> 2523    make[5]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:3770: tree] Err
             or 1
     2524    make[5]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructur
             e/IO'
  >> 2525    make[4]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:3770: tree] Err
             or 1
     2526    make[4]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src/Infrastructur
             e'
  >> 2527    make[3]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:3770: tree] Err
             or 1
     2528    make[3]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/src'
  >> 2529    make[2]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:3770: tree] Err
             or 1
     2530    make[2]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src'
  >> 2531    make[1]: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:2245: build_lib
             s] Error 2
     2532    make[1]: Leaving directory '/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src'
  >> 2533    make: *** [/home/ubuntu/sandpit/spack-stack-1.0.0-rc1/cache/build_stage/spack-stage-esmf-8.3.0-z35xze4xsc7y4agphvhe3vbdns3rqhor/spack-src/build/common.mk:2237: lib] Error 2
```

### Information on your system

```
[skylab-1.0.0] ubuntu@ip-172-31-79-27:~/sandpit/spack-stack-1.0.0-rc1$ spack debug report
* **Spack:** 0.18.0 (cb1377b6117220e40e7d454e7c5b94e44f20ca53)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-haswell
* **Concretizer:** clingo
```

### Additional information

Workaround: install cmake via `apt install` on the AWS Ubuntu image.

Solution: Fix esmf package in spack fork.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by jedwards4b**
For ESMF 8.3.0 and newer use external parallelio build with parallelio 2.5.8 or newer.
Add esmf develop branch as an option.   "develop" has a special meaning in spack and is considered newer than any named tag.

### CODE REVIEW

**Review by climbfuji — COMMENTED** [review]
I see we have different ideas about external/internal PIO. Our (@kgerheiser and my) thinking was that ESMF versions that are able to use an external PIO should be hardcoded to do so, in order to avoid inconsistencies in the software stack (two versions of PIO being built and used, one by spack directly and one within ESMF).

Only for older ESMF versions would we allow building PIO internally, since there is no other way.

What's the reasoning behind allowing internal PIO builds for newer ESMF versions that are able to use a spack-built one?

**Comment by jedwards4b** [comment]
@climbfuji That is what I did - the 8.3.0 and newer will always use an external parallelio build (spack reviewers requested this option name rather than pio) 
and older versions use internal pio.   The key is in lines 43 and 45.

**Comment by climbfuji** [comment]
> 

Not exactly. You still allow turning off parallelio for 8.3.0+ and turning it on for earlier versions.  You could fix that by adding `conflict` lines, I believe.

Also, do we need to hardcode the `pnetcdf` option to disable the Fortran API? I think ESMF would build fine using a pnetcdf installation that has the Fortran API enabled. With the dependency spec as in this PR, I think it would build a second parallel-netcdf package if the existing one has the Fortran API enabled. Not entirely sure, but that is how I think it works.

**Comment by jedwards4b** [comment]
Turning off parallelio for esmf > 8.3.0 is a legitimate option - ESMF_PIO=OFF.
The parallelio option is not offered for versions older than 8.3.0 due to 
       when="@8.3:",        at line 43.  

No we do not need to hardcode the pnetcdf option to disable fortran and I have not done that.   If the pnetcdf option is provided to esmf then we want a parallelio build with pnetcdf also.   These options refer to the parallelio build and not the pnetcdf build.   We don't ever need the fortran library of parallelio for esmf.

**Comment by climbfuji** [comment]
> Turning off parallelio for esmf > 8.3.0 is a legitimate option - ESMF_PIO=OFF. The parallelio option is not offered for versions older than 8.3.0 due to when="@8.3:", at line 43.
> 
> No we do not need to hardcode the pnetcdf option to disable fortran and I have not done that. If the pnetcdf option is provided to esmf then we want a parallelio build with pnetcdf also. These options refer to the parallelio build and not the pnetcdf build. We don't ever need the fortran library of parallelio for esmf.

Slowly getting it ... too much sunshine and high altitude during my vacation last week.

Still wondering if lines 59 and 60 will explicitly request parallelio without the Fortran API instead of using one that has it enabled (even though ESMF would be fine with an external parallelio that has the Fortran API enabled).

**Comment by uturuncoglu** [comment]
@jedwards4b @climbfuji Just note that default for xerces is still True in here. I think it must be False by default. If you want to use XML output of attributes then you could activate it.

**Comment by jedwards4b** [comment]
The failing test is due to parallelio 2.5.8 not being available on this branch - it is on spack/spack develop but a merge gives too many conflicts.   Not sure how to proceed.

I am digging into the parallelio-fortran issue a bit more.   We want esmf to use parallelio with fortran if it is already there but if it is to be built just for esmf then we do not want the fortran variant.

**Comment by climbfuji** [comment]
@kgerheiser was going to bring over the formatting-overhaul PR from Todd from the authoritative repo. If he didn't get a chance to do this yet, I can take care of it and then bring over the missing parallelio changes.

**Review by climbfuji — COMMENTED** [review]
Basically ready to merge, but blocked by the missing parallelio 2.5.8 version. Will approve once we got parallelio 2.5.8 in the branch.

**Comment by climbfuji** [comment]
@jedwards4b I am working on the update of our branch from the authoritative repository, see https://github.com/NOAA-EMC/spack/pull/153. Once that is merged, the necessary PIO updates will be there.

**Comment by climbfuji** [comment]
@jedwards4b I merged the update the from the authoritative spack repository develop branch this morning. It's now showing conflicts in your ESMF package PR, but you should be able to resolve those easily (all of the formatting changes are coming from spack develop due to their new style checker `black`). At least, you know have the latest PIO updates and what not that you need for your ESMF PR.

**Comment by jedwards4b** [comment]
@climbfuji Apparently I cannot resolve the conflicts, please accept incoming changes in all cases.

**Comment by climbfuji** [comment]
> @climbfuji Apparently I cannot resolve the conflicts, please accept incoming changes in all cases.

Let me try to fix those for you.

**Comment by climbfuji** [comment]
I'll get this done this morning after my first round of meetings.

**Comment by climbfuji** [comment]
@jedwards4b FYI - https://github.com/NOAA-EMC/spack/pull/154

**Comment by jedwards4b** [comment]
@climbfuji I've merged in #154 and fixed the black issues.   Can you start the workflow?

**Comment by climbfuji** [comment]
> @climbfuji I've merged in #154 and fixed the black issues. Can you start the workflow?

It's running now

**Review by climbfuji — APPROVED** [review]

**Comment by climbfuji** [comment]
@jedwards4b Thanks for that PR. I'll send those changes back to the authoritative spack repo next.

**Pull request merged**
